scripts/run_tests.sh

```diff
@@ -0,0 +1,447 @@
+
+    log "Cleanup completed." $COLOR_GREEN
+}
+
+# Set trap to ensure cleanup on script exit (normal or abnormal)
+trap cleanup EXIT
+
+# Copy project files to the temporary directory
+log "Copying project files to temporary directory..." $COLOR_CYAN
+
+# Compress the project directory
+log "Compressing project files..." $COLOR_CYAN
+(cd .. && tar -czf "$TEST_DIR/rengine-project.tar.gz" --exclude='docker/secrets' .)
+
+cd "$TEST_DIR"
+
+# Download appropriate Debian 12 cloud image
+log "Downloading Debian 12 cloud image for $ARCH..." $COLOR_CYAN
+if [ "$ARCH" = "amd64" ]; then
+    wget -q https://cloud.debian.org/images/cloud/bookworm/latest/debian-12-generic-amd64.qcow2 -O debian-12-generic.qcow2
+elif [ "$ARCH" = "arm64" ]; then
+    wget -q https://cloud.debian.org/images/cloud/bookworm/latest/debian-12-generic-arm64.qcow2 -O debian-12-generic.qcow2
+fi
+
+# Create a larger disk image
+qemu-img create -f qcow2 -o preallocation=metadata "$TEST_DIR/large-debian.qcow2" $VM_DISK_SIZE
+
+# Resize the downloaded image to fill the new larger disk
+qemu-img resize "$TEST_DIR/debian-12-generic.qcow2" $VM_DISK_SIZE
+
+# Combine the two images
+qemu-img convert -O qcow2 -o preallocation=metadata "$TEST_DIR/debian-12-generic.qcow2" "$TEST_DIR/large-debian.qcow2"
+
+# Create a copy of the image for testing
+mv large-debian.qcow2 test-debian.qcow2
+
+# Generate SSH key pair
+log "Generating SSH key pair..." $COLOR_CYAN
+ssh-keygen -t rsa -b 2048 -f ./id_rsa -N ""
+
+# Create a cloud-init configuration file
+cat > cloud-init.yml <<EOF
+#cloud-config
+users:
+  - name: rengine
+    sudo: ALL=(ALL) NOPASSWD:ALL
+    shell: /bin/bash
+    ssh_authorized_keys:
+      - $(cat ./id_rsa.pub)
+EOF
+
+# Create a cloud-init ISO
+cloud-localds cloud-init.iso cloud-init.yml
+
+# Start the VM
+log "Starting the VM..." $COLOR_CYAN
+if [ "$ARCH" = "amd64" ]; then
+    qemu-system-x86_64 \
+        -name $VM_NAME \
+        -m $VM_RAM \
+        -smp $VM_CPUS \
+        -enable-kvm \
+        -cpu host \
+        -nodefaults \
+        -no-fd-bootchk \
+        -drive file=$VM_IMAGE,format=qcow2 \
+        -drive file=cloud-init.iso,format=raw \
+        -device virtio-net-pci,netdev=net0 \
+        -netdev user,id=net0,hostfwd=tcp::2222-:22,hostfwd=tcp::8443-:443 \
+        -vga std \
+        -vnc :0 \
+        -display none &
+elif [ "$ARCH" = "arm64" ]; then
+    qemu-system-aarch64 \
+        -name $VM_NAME \
+        -M virt \
+        -m $VM_RAM \
+        -smp $VM_CPUS \
+        -cpu cortex-a72 \
+        -nodefaults \
+        -drive file=$VM_IMAGE,format=qcow2 \
+        -drive file=cloud-init.iso,format=raw \
+        -device virtio-net-pci,netdev=net0 \
+        -netdev user,id=net0,hostfwd=tcp::2222-:22,hostfwd=tcp::8443-:443 \
+        -device virtio-gpu-pci \
+        -device ramfb \
+        -device nec-usb-xhci,id=xhci \
+        -device usb-kbd \
+        -device usb-tablet \
+        -vnc :0 \
+        -serial mon:stdio \
+        -display none &
+fi
```

Comment:
I will restrict this to only WSL user
Running script with sudo is not needed

Reply:
Done could you test please

Reply:
I think you need to place the check at the very beginning of the script, because now I have to download packages, download Debian etc. before I get a 'permission denied':

![image](https://github.com/user-attachments/assets/39b1f3b5-8ec8-48fc-a602-ed9529e9fd8a)

Due to that, script doesn't tell me I need to execute this as `sudo`.

Reply:
Have you the latest version of the script ?
docker/secrets should have been avoided
accorded to this
  # Compress the project directory
    log "Compressing project files..." $COLOR_CYAN
    (cd .. && tar -czf "$TEST_DIR/rengine-project.tar.gz" --exclude='docker/secrets' .)

Reply:
My bad I put docker/secrets instead of docker/certs. I made a mistake

Reply:
I did a test with docker/secrets/certs as root user and it's excluded from the tar
And no permission denied